### PR TITLE
[MUGEN] Add axial attention blocks for encoder and decoder

### PR DESCRIPTION
**Pull request by RdoubleA**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101
* #98
* #87
* __->__ #86

## Summary
Refactor `AxialBlock` and `AttentionResidualBlock` from [MUGEN](https://github.com/mugen-org/MUGEN_baseline/blob/main/lib/models/video_vqvae/vqvae.py). Included some minor edits to `MultiHeadAttention` after testing attention blocks. `AxialBlock` was placed in `attention.py` since it can be generally used to apply axial attention to n-dimensional inputs. `AttentionResidualBlock` was placed in `video_vqvae.py` since it is specifically used by that model and is not yet generalized.

## Test Plan
Compare output to MUGEN's code directly and run added unit tests.

`AxialBlock`
image

`AttentionResidualBlock`
image

```
test/modules/layers/test_attention.py::TestAttention::test_axial_attention PASSED                                             [ 57%]
test/modules/layers/test_attention.py::TestAttention::test_axial_block_channel_dim PASSED                                     [ 58%]
test/modules/layers/test_attention.py::TestAttention::test_axial_block_forward PASSED                                         [ 59%]
test/models/test_video_vqvae.py::TestAttentionResidualBlock::test_hidden_dim_assertion PASSED                                 [ 14%]
test/models/test_video_vqvae.py::TestAttentionResidualBlock::test_forward PASSED                                              [ 15%]
```

Differential Revision: [D37288326](https://our.internmc.facebook.com/intern/diff/D37288326)

### CODE REVIEW

File: test/modules/layers/test_attention.py [hunk 1]
```diff
@@ -125,22 +129,56 @@ def test_combine_multihead(self):
 
     def test_multi_head_attention(self):
         # New tensors because need unflattened shape
-        q = torch.randn(1, *self.input_shape, self.hidden_dim)
-        k = torch.randn(1, *self.input_shape, self.hidden_dim)
-        v = torch.randn(1, *self.input_shape, self.hidden_dim)
+        q = 2 * torch.ones(1, *self.input_shape, self.hidden_dim)
```
Comment (ankitade) [inline test/modules/layers/test_attention.py:132]: is this change related?

File: test/modules/layers/test_attention.py [hunk 2]
```diff
@@ -125,22 +129,56 @@ def test_combine_multihead(self):
...
+                        [[2.4187, 4.1634, 1.4579], [2.4187, 4.1634, 1.4579]],
                     ],
                 ]
             ]
         )
         assert_expected(actual, expected, rtol=0, atol=1e-4)
+
+    def test_axial_block_mha_length(self):
+        """Test AxialAttentionBlock number of MHAs"""
+        assert len(self.block.mha_attns) == 3, "incorrect number of MHAs"
```
Comment (ebsmothers) [inline test/modules/layers/test_attention.py:154]: Do we really need this test? Feels like it is just testing that the test initialization is working as expected

File: torchmultimodal/models/video_vqvae.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    def __init__(self, hidden_dim: int, n_head: int = 2) -> None:
+        super().__init__()
+        # To avoid hidden dim becoming 0 in middle layers
+        if hidden_dim < 2:
+            raise ValueError("hidden dim must be at least 2")
+
+        self.block = nn.Sequential(
+            nn.BatchNorm3d(hidden_dim),
+            nn.ReLU(),
+            SamePadConv3d(hidden_dim, hidden_dim // 2, 3, bias=False),
```
Comment (ebsmothers) [inline torchmultimodal/models/video_vqvae.py:37]: I know the downsampling comes from the VideoVQVAE implementation, but wondering if there is any case users would want to downsample with a different frequency within the attention residual block

File: torchmultimodal/modules/layers/attention.py [hunk 4]
```diff
@@ -12,6 +13,67 @@
...
+    """
+
+    def __init__(self, n_dims: int, qkv_dim: int, n_head: int) -> None:
+        super().__init__()
+        self.qkv_dim = qkv_dim
+        self.mha_attns = nn.ModuleList(
+            [
+                MultiHeadAttention(
+                    shape=tuple(
+                        repeat(0, n_dims)
```
Comment (ebsmothers) [inline torchmultimodal/modules/layers/attention.py:48]: Can you instead make shape optional in MHA and assert that it gets passed whenever causal is true?

File: torchmultimodal/modules/layers/attention.py [hunk 5]
```diff
@@ -12,6 +13,67 @@
...
+                    ),  # dummy value for shape since we are not using causal
+                    dim_q=qkv_dim,
+                    dim_kv=qkv_dim,
+                    n_head=n_head,
+                    n_layer=1,
+                    causal=False,
+                    attn_module=AxialAttention(d),
```
Comment (ebsmothers) [inline torchmultimodal/modules/layers/attention.py:55]: Not necessarily a comment for this PR, but it might be useful to provide more info on the in/out dims expected for attn_module in the docstring for MultiHeadAttention. Though your usage here provides a good example, being explicit about the APIs will make it even easier for any users that want to plug in their own attention modules.

File: torchmultimodal/modules/layers/attention.py [hunk 6]
```diff
@@ -12,6 +13,67 @@
...
+    https://github.com/wilson1yan/VideoGPT/blob/master/videogpt/vqvae.py
+
+    Args:
+        n_dims (int): dimensionality of input data, not including batch or channel dims
+        qkv_dim (int): dimensionality of linear projections Wq, Wk, and Wv in attention
+        n_head (int): number of heads in multihead attention. Must divide into qkv_dim
+                      evenly
+
+    Inputs:
+        x (Tensor): a [b, c, d1, ..., dn] tensor, where c == qkv_dim
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply torchmultimodal/modules/layers/attention.py:38]: a is not a variable here, just referring to "a tensor". b is batch. I can edit to just say [b,c,d,...] tensor

File: test/modules/layers/test_attention.py (hunk 1 above)
Comment (RdoubleA) [reply test/modules/layers/test_attention.py:132]: not directly, but since I changed the setUp that affected all the random tensors since they generate different values based on the order they are called... I figured it was too unstable so I made it a fixed input instead.

File: torchmultimodal/models/video_vqvae.py (hunk 3 above)
Comment (RdoubleA) [reply torchmultimodal/models/video_vqvae.py:37]: I am not aware of any use case of downsampling differently within the block. I am leaning towards not having this be configurable because 1) it is downsampled down and back up within the block, 2) downsampling should occur in the conv layers before this block, 3) this block is already video vqvae specific so I don't see the point of making it configurable. WDYT?

File: torchmultimodal/modules/layers/attention.py (hunk 4 above)
Comment (RdoubleA) [reply torchmultimodal/modules/layers/attention.py:48]: good idea

File: torchmultimodal/modules/layers/attention.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (RdoubleA) [reply torchmultimodal/modules/layers/attention.py:38]: actually I will keep it otherwise I have to edit all the docstrings in this file to be consistent

**Comment by RdoubleA** [comment]
@ebsmothers I will probably make the changes you suggested to `MultiHeadAttention` in a separate PR

**Review by ebsmothers — APPROVED** [review]
Looks good to me, just make sure to address the MHA stuff in a subsequent PR

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37288326).
